### PR TITLE
Test specialization's thread-safety

### PR DESCRIPTION
**Pull request by brandtbucher**

### CODE REVIEW

File: Lib/test/test_opcache.py [hunk 1]
```diff
@@ -481,6 +484,513 @@ def f(x, y):
             f()
 
 
+class TestRacesDoNotCrash(unittest.TestCase):
+    # Careful with these. Bigger numbers have a higher chance of catching bugs,
+    # but you can also burn through a *ton* of type/dict/function versions:
+    ITEMS = 1 << 10
```
Comment (markshannon) [inline Lib/test/test_opcache.py]: Any reason why these need to be powers of 2? If not, maybe use `1000` instead.

File: Lib/test/test_opcache.py [hunk 2]
```diff
@@ -481,6 +484,513 @@ def f(x, y):
...
+                self.assert_specialized(read, opname)
+            # Create writers:
+            writers = []
+            for _ in range(self.WRITERS):
+                writer = threading.Thread(target=write, args=[items])
+                writers.append(writer)
+            # Run:
+            for writer in writers:
+                writer.start()
+            read(items)  # BOOM!
```
Comment (markshannon) [inline Lib/test/test_opcache.py:528]: Would increasing the number of reader threads allow us to burn through fewer versions by increasing the chance of failure per version change?

**Review by markshannon — APPROVED** [review]
LGTM.
A couple of comments, but nothing showstopping.

File: Lib/test/test_opcache.py (hunk 1 above)
Comment (brandtbucher) [reply Lib/test/test_opcache.py]: It was useful when I was trying to dial in a good value while being aware of exactly how many versions I was using. I can change them, though.

File: Lib/test/test_opcache.py (hunk 2 above)
Comment (brandtbucher) [reply Lib/test/test_opcache.py:528]: I didn't see a significant impact on crashiness when dialing the thread count up to moderate levels, and beyond that it just took *forever* to run.

**Pull request merged**